internal/repository/repository.go

```diff
@@ -252,12 +322,51 @@ func (r *Repository) LookupBlobSize(id restic.ID, tpe restic.BlobType) (uint, bo
 	return r.idx.LookupSize(restic.BlobHandle{ID: id, Type: tpe})
 }
 
+func (r *Repository) getZstdEncoder() *zstd.Encoder {
+	r.allocEnc.Do(func() {
+		level := zstd.SpeedDefault
+		if r.opts.Compression == CompressionMax {
+			level = zstd.SpeedBestCompression
+		}
+
+		enc, err := zstd.NewWriter(nil, zstd.WithEncoderLevel(level))
+		if err != nil {
+			panic(err)
+		}
+		r.enc = enc
+	})
+	return r.enc
+}
+
+func (r *Repository) getZstdDecoder() *zstd.Decoder {
+	r.allocDec.Do(func() {
+		dec, err := zstd.NewReader(nil)
+		if err != nil {
+			panic(err)
+		}
+		r.dec = dec
+	})
+	return r.dec
+}
+
 // saveAndEncrypt encrypts data and stores it to the backend as type t. If data
 // is small enough, it will be packed together with other small blobs.
 // The caller must ensure that the id matches the data.
 func (r *Repository) saveAndEncrypt(ctx context.Context, t restic.BlobType, data []byte, id restic.ID) error {
 	debug.Log("save id %v (%v, %d bytes)", id, t, len(data))
 
+	uncompressedLength := 0
+	if r.cfg.Version > 1 {
+
+		// we have a repo v2, so compression is available. if the user opts to
+		// not compress, we won't compress any data, but everything else is
+		// compressed.
+		if r.opts.Compression != CompressionOff || t != restic.DataBlob {
+			uncompressedLength = len(data)
+			data = r.getZstdEncoder().EncodeAll(data, nil)
+		}
```

Comment:
Since you store information about each block, we might as well just skip storing bigger or equal blocks.

Incompressible block will have a frame and block header, so it will be a few bytes bigger. Also, this reduces the JSON size and removes any need for decompression copying when reading:

```suggestion
		if r.opts.Compression != CompressionOff || t != restic.DataBlob {
			compressed = r.getZstdEncoder().EncodeAll(data, nil)
			if len(compressed) < len(data) {
				uncompressedLength = len(data)
				data = compressed
			} 
		}
```

Reply:
Good idea in general, but we'd like to make sure that all tree blobs (JSON documents) are compressed, so we can require that for a later version of the repository format. The code here should always compress all JSON documents (because they compress very well), except maybe the empty tree (`{}`).